File: observer/java/ObserverPatternDemo.java
import java.util.Vector;
import java.util.List;

interface Observer {

	<T> void update(T state);
}

abstract class Publisher {

	private String state;
	private final List<Observer> observers;

	protected Publisher() {
		this.observers = new Vector();
		this.state = "";
	}

	public void setState(String state) {
		this.state = state;
		this.notifyAllObservers();
	}

	void subscribe(Observer observer) {
		this.observers.add(observer);
	}

	void unsubscribe(Observer observer) {
		this.observers.remove(observer);
	}

 	private void notifyAllObservers() {
		for (Observer o : this.observers) o.<String>update(this.state);
	}
}


final class ConcretePublisher extends Publisher {

	public ConcretePublisher() {
		super();
	}
}


final class ConcreteObserver implements Observer {

	private String name;

	public ConcreteObserver(String name) {
		this.name = name;
	}

	@Override
	public <T> void update(T state) {
		System.out.printf("[State notified] name: %s, state: %s\n", this.name, state+"");
	}
}

final class ObserverPatternDemo {

	public static void main(String[] args) {
		Publisher pub = new ConcretePublisher();

		Observer obs1 = new ConcreteObserver("obs1");
		Observer obs2 = new ConcreteObserver("obs2");
		Observer obs3 = new ConcreteObserver("obs3");
		Observer obs4 = new ConcreteObserver("obs4");

		pub.subscribe(obs1);
		pub.subscribe(obs2);
		pub.subscribe(obs3);
		pub.subscribe(obs4);
		pub.setState("state 1");
		pub.setState("state 4");
		pub.setState("state 7");
		pub.setState("state 10");
	}
}

File: observer/kotlin/observer_pattern_demo.kt
interface Observer {

	fun <T> update(state: T): Unit
}

abstract class Publisher(private val observers: MutableList<Observer>) {

	var state: String = ""
		set(value) = this.notifyAllObservers()

	protected constructor() : this(mutableListOf())

	fun subscribe(observer: Observer): Unit {
		this.observers.add(observer)
	}

	fun unsubscribe(observer: Observer): Unit {
		this.observers.remove(observer)
	}

	private fun notifyAllObservers(): Unit {
		this.observers.forEach { it.update<String>(this.state) }
	}
}

class ConcretePublisher : Publisher()

class ConcreteObserver(private val name: String) : Observer {

	override fun <T> update(state: T): Unit {
		println("[State notified] name: ${this.name}, state: $state")
	}
}

fun main(): Unit {
	val pub: Publisher = ConcretePublisher()
	val obs1: Observer = ConcreteObserver("obs1");
	val obs2: Observer = ConcreteObserver("obs2");
	val obs3: Observer = ConcreteObserver("obs3");
	val obs4: Observer = ConcreteObserver("obs4");
	val obs5: Observer = ConcreteObserver("obs5");

	pub.subscribe(obs1)
	pub.subscribe(obs2)
	pub.subscribe(obs3)
	pub.subscribe(obs4)
	pub.subscribe(obs5)

	pub.state = "State 1"
	pub.state = "State 2"
	pub.state = "State 3"
}